spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -174,41 +187,74 @@ protected Object build(Object payload, MessageHeaders headers) {
+			}
+			else if (this.mimeType.equals(MessageConverterUtils.X_SPRING_TUPLE)) {
+				this.klazz = Tuple.class;
+			}
+			else if (this.mimeType.getType().equals("text") || this.mimeType.getSubtype().equals(
+					"json") || this.mimeType.getSubtype().equals("xml")) {
+				this.klazz = String.class;
+			}
+			else {
+				this.klazz = byte[].class;
+			}
+			this.messageConverter = new MessageWrappingMessageConverter(
+						MessageConverterConfigurer.this.compositeMessageConverterFactory
+								.getMessageConverterForType(this.mimeType), this.mimeType);
 		}
 
 		@Override
 		public Message<?> preSend(Message<?> message, MessageChannel channel) {
-			Class<?>[] classes =
-					MessageConverterConfigurer.this.compositeMessageConverterFactory.supportedDataTypes(
-							this.mimeType);
-			MessageWrappingMessageConverter messageConverter =
-					new MessageWrappingMessageConverter(
-							MessageConverterConfigurer.this.compositeMessageConverterFactory
-									.getMessageConverterForType(this.mimeType), this.mimeType);
-			for (Class<?> aClass : classes) {
-				if (aClass.isAssignableFrom(message.getPayload().getClass())) {
-					Object contentTypeFromMessage = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
-					if (contentTypeFromMessage == null) {
-						return MessageConverterConfigurer.this.messageBuilderFactory
-								.fromMessage(message)
-								.setHeader(MessageHeaders.CONTENT_TYPE, this.contentType)
-								.build();
-					}
-					else {
-						return message;
-					}
+			Message<?> sentMessage = null;
+			if (this.klazz.isAssignableFrom(message.getPayload().getClass())) {
+				Object contentTypeFromMessage = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
+				if (contentTypeFromMessage == null) {
+					sentMessage = MessageConverterConfigurer.this.messageBuilderFactory
+							.fromMessage(message)
+							.setHeader(MessageHeaders.CONTENT_TYPE, this.contentType)
+							.build();
 				}
 				else {
-					Object converted = messageConverter.fromMessage(message, aClass);
-					if (converted != null) {
-						return (Message<?>) converted;
+					sentMessage = message;
+				}
+			}
+			else {
+				Object converted = this.input ? this.messageConverter.fromMessage(message, this.klazz)
+						: this.messageConverter.toMessage(message.getPayload(), message.getHeaders(), null);
+				if (converted != null) {
+					Message<?> convertedMessage = (Message<?>) converted;
+					Object contentType = convertedMessage.getHeaders().get(MessageHeaders.CONTENT_TYPE);
+					if (contentType instanceof MimeType) {
+						sentMessage = MessageConverterConfigurer.this.messageBuilderFactory
+								.fromMessage(convertedMessage).setHeader(
+										MessageHeaders.CONTENT_TYPE,
+										contentType.toString()).build();
+					}
+					else {
+						sentMessage = convertedMessage;
 					}
 				}
 			}
-			throw new MessageConversionException("Cannot convert " + message + " to " + this.contentType);
+			if (sentMessage != null) {
```

Comment:
I think this would read better as:

```
if (sentMessage == null) {
    throw new MessageConversionException("Cannot convert " + message + " to " + this.contentType);
 }
return sentMessage;
```